calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -56,78 +83,213 @@ type ResourceActionCommand func(*client.Client, unversioned.Resource) (unversion
+	isNamespaced      bool
 	create            ResourceActionCommand
 	update            ResourceActionCommand
 	delete            ResourceActionCommand
+	get               ResourceActionCommand
 	list              ResourceActionCommand
 }
 
 func (r resourceHelper) String() string {
-	return fmt.Sprintf("Resource(%s %s)", r.typeMetadata.Kind, r.typeMetadata.APIVersion)
+	if !r.isList {
+		return fmt.Sprintf("Resource(%s %s)", r.resource.GetObjectKind(), r.resource.GetObjectKind().GroupVersionKind())
+
+	}
+	return fmt.Sprintf("Resource(%s %s)", r.listResource.GetObjectKind(), r.listResource.GetListMeta().GetResourceVersion())
 }
 
-// Store a resourceHelper for each resource unversioned.TypeMetadata.
-var helpers map[unversioned.TypeMetadata]resourceHelper
+// Store a resourceHelper for each resource.
+var helpers map[schema.GroupVersionKind]resourceHelper
+var kindToRes = make(map[string]runtime.Object)
 
-func registerResource(res unversioned.Resource, resList unversioned.Resource,
+func registerResource(res runtime.Object, resList runtime.Object, isNamespaced bool, names []string,
 	tableHeadings []string, tableHeadingsWide []string, headingsMap map[string]string,
-	apply, create, update, delete, list ResourceActionCommand) {
+	create, update, delete, get, list ResourceActionCommand) {
 
 	if helpers == nil {
-		helpers = make(map[unversioned.TypeMetadata]resourceHelper)
+		helpers = make(map[schema.GroupVersionKind]resourceHelper)
 	}
 
-	tmd := res.GetTypeMetadata()
 	rh := resourceHelper{
-		typeMetadata:      tmd,
+		resource:          res,
 		resourceType:      reflect.ValueOf(res).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            false,
-		apply:             apply,
+		isNamespaced:      isNamespaced,
 		create:            create,
 		update:            update,
 		delete:            delete,
+		get:               get,
 		list:              list,
 	}
-	helpers[tmd] = rh
+	helpers[res.GetObjectKind().GroupVersionKind()] = rh
 
-	tmd = resList.GetTypeMetadata()
 	rh = resourceHelper{
-		typeMetadata:      tmd,
+		listResource:      resList.(ResourceListObject),
 		resourceType:      reflect.ValueOf(resList).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            true,
 	}
-	helpers[tmd] = rh
+	helpers[resList.GetObjectKind().GroupVersionKind()] = rh
+
+	for _, v := range names {
+		kindToRes[v] = res
+	}
+}
+
+func (rh resourceHelper) GetObjectType() reflect.Type {
+	return rh.resourceType
+}
+
+// Apply is an un-typed method to apply (create or update) a resource. This calls Create
+// and if the resource already exists then we call the Update method.
+func (rh resourceHelper) Apply(client client.Interface, resource runtime.Object) (runtime.Object, error) {
+
+	// Try to assert the resource type and remove the resourceVersion, because
```

Comment:
Did you ascertain what the correct behavior should be for apply?  should it care about the ResourceVersion or not?

Reply:
yes they care about it, so what I'm thinking is we should do an update first (which cares about the resourceVersion) and if that fails then we do a Create (which ignores it)